src/chaat/handler/handler.clj

```diff
@@ -0,0 +1,52 @@
+(ns chaat.handler.handler
+  (:require [ring.util.response :as res]
+            [chaat.model.user :as model.user]
+            [java-time.api :as jt]
+            [chaat.handler.validation :as handler.validation]
+            [chaat.errors :refer [do-or-error]]
+            [chaat.db.utils :as db.utils]))
+
+(defn home
+  "Respond with a greeting/welcome"
+  [request]
+  (res/response "Welcome to chaat"))
+
+(defn health-check
+  "Respond with application status: checks server and db status"
+  [db request]
+  (let [result (db.utils/health-check db)]
+    (if (nil? (:error result))
+      (res/response (str "all systems go: " (jt/instant)))
+      (res/bad-request (str (:error result))))))
+
+(defn signup
+  "Create a user account with supplied parameters"
+  [db request]
+  (let [params (:params request)
+        {:keys [username password]} params
+        result (handler.validation/validate-signup-details username password)
+        result (do-or-error result model.user/create db username password)]
+    (if (nil? (:error result))
+      (res/response "Signup successful")
+      (res/bad-request (str (:error result))))))
```

Comment:
[bug]
There is no differentiation between a client error and a server error here. This is true of the other handler functions in this namespace and they need to be fixed as well.

If the signup detail validation failed, return Bad Request (400). If the error happened during the creation of the user model in line 27, return Internal Server Error (500).

In general, have a look at the possible options for HTTP statuses in MDN, and return the appropriate one.

Reply:
Fixed this. We are now returning the correct status codes, and I'll improve this code iteratively.